components/server/src/main/java/com/hotels/styx/server/netty/codec/NettyToStyxRequestDecoder.java

```diff
@@ -63,85 +63,103 @@
                 out.add(styxRequest);
-            } else if (httpObject instanceof HttpContent && this.producer != null) {
-                this.producer.onNext(content(httpObject));
 
-                if (httpObject instanceof LastHttpContent) {
-                    this.producer.onCompleted();
+            }
+            if (msg instanceof HttpContent) {
+                ByteBuf content = ((ByteBufHolder) msg).content();
+                if (content.isReadable()) {
+                    getContentProducer(ctx).newChunk(retain(content));
+                }
+                if (msg instanceof LastHttpContent) {
+                    getContentProducer(ctx).lastHttpContent();
                 }
             }
         } catch (BadRequestException ex) {
             throw ex;
         } catch (Exception ex) {
-            throw new BadRequestException(ex.getMessage() + " in " + httpObject, ex);
+            throw new BadRequestException(ex.getMessage() + " in " + msg, ex);
         }
     }
 
-    @Override
-    public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
-        if (this.producer != null) {
-            this.producer.notifySubscriber();
-        }
-        super.channelReadComplete(ctx);
-    }
-
     @Override
     public void channelInactive(ChannelHandlerContext ctx) throws Exception {
-        cleanUp();
+        TransportException cause = new TransportException(
+                "Connection to client lost: " + ctx.channel().remoteAddress());
+        getContentProducer(ctx).channelInactive(cause);
         super.channelInactive(ctx);
     }
 
     @Override
     public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
-        cleanUp();
-        if (cause instanceof TooLongFrameException) {
-            throw new BadRequestException(cause.getMessage(), cause);
-        }
+        getContentProducer(ctx).channelException(toStyxException(cause));
         super.exceptionCaught(ctx, cause);
     }
 
-    private void cleanUp() {
-        if (producer != null) {
-            producer.cleanUp();
+    private FlowControllingHttpContentProducer getContentProducer(ChannelHandlerContext ctx) {
+        if (!this.producer.isPresent()) {
+            this.producer = Optional.of(createProducer(ctx));
+        }
+        return this.producer.get();
+    }
+
+    private FlowControllingHttpContentProducer createProducer(ChannelHandlerContext ctx) {
+        String loggingPrefix = format("%s -> %s", ctx.channel().remoteAddress(), ctx.channel().localAddress());
+
+        return new FlowControllingHttpContentProducer(
+                () -> ctx.channel().read(),
+                () -> ctx.channel().config().setAutoRead(true),
+                cause -> { },
+                format("%s, %s", loggingPrefix, ""),
+                null,
+                inactivityTimeoutMs);
+    }
+
+    private Throwable toStyxException(Throwable cause) {
```

Comment:
can we use a more meaningful name here? It seems other kinds of Netty exceptions could be caught here but we are only dealing with TooLongFrameException (I don't know why!)